File: autoviz/examples/cpp/22_tutorial_service.cpp
/******************************************************************************
 * Copyright 2026 The Openbot Authors (duyongquan)
 *****************************************************************************/

#include <iostream>
#include <memory>
#include <string>

#include <automsgs/msgs/std_msgs/string.pb.h>
#include <unistd.h>

#include "common/tutorial_utils.hpp"

using String = automsgs::msgs::std_msgs::String;

int main(int argc, char** argv) {
  if (!autolink::Init(argv[0])) return 1;
  auto node = autolink::CreateNode("/autoviz/service");
  auto server = node->CreateService<String, String>(
      "/fake/echo",
      [](const std::shared_ptr<String>& req, std::shared_ptr<String>& res) {
        std::cout << "echo ← " << req->data() << "\n";
        res->set_data("echo:" + req->data());
      });
  if (server == nullptr) {
    std::cerr << "failed to create /fake/echo service\n";
    return 1;
  }
  std::cout << "Service /fake/echo ready — use Autoviz Service Call panel\n";

  while (autolink::OK()) {
    usleep(200000);
  }
  return 0;
}

File: autoviz/examples/python/22_tutorial_service.py
#!/usr/bin/env python3
"""Host /fake/echo String service for Autoviz Service Call panel.

  Service: /fake/echo  std_msgs/String → echo request

  /usr/bin/python3 examples/python/22_tutorial_service.py
  ./build/autonomy/bin/autoviz
"""

from __future__ import annotations

from _reexec import early_reexec_if_needed

early_reexec_if_needed()

import time

from _bootstrap import prepare_example_environment

prepare_example_environment()

import autolink
from automsgs.msgs.std_msgs import string_pb2


def echo(req):
    print(f'echo ← {req.data!r}')
    return string_pb2.String(data=f'echo:{req.data}')


def main():
    autolink.init('service')
    n = autolink.Node('/autoviz/service')
    n.create_service('/fake/echo', echo, req_type=string_pb2.String)
    print('Service /fake/echo ready — use Autoviz Service Call panel')
    try:
        while not autolink.is_shutdown():
            time.sleep(0.2)
    except KeyboardInterrupt:
        pass
    finally:
        autolink.shutdown()


if __name__ == '__main__':
    main()
